### PR TITLE
feat: Support npm workspaces caching

### LINKED ISSUE #751: Improve caching for NPM Workspaces
**Is your feature request related to a problem? Please describe.**

NPM Workspaces as of version 8 are now context aware.

This means that when using npm 8 in a mono repo and Netlify you can safely run `npm i` from with a folder and it will respect the workspace and install properly.

The Netlify build image however doesn't support this as it does not cache the root level node modules and on subsequent runs will fail to build the site.

 
**Describe the solution you'd like**

To fix this the build-image will need to be updated to handle npm workspaces caching similarly to yarn.

**Describe alternatives you've considered**

N/A

**Additional context**

**Can you submit a pull request?**

Yes

> Pull requests are welcome! If you would like to help us add this feature, please check our
> [contributions guidelines](../../CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by zackerydev**
🎉 Thanks for submitting a pull request! 🎉

#### Summary

Fixes #751 and related to #196

---

For us to review and ship your PR efficiently, please perform the following steps:

- [x] Open a [bug/issue](https://github.com/netlify/build-image/issues/new/choose) before writing your code 🧑‍💻. This ensures we can discuss the changes and get feedback from everyone that should be involved. If you\`re fixing a typo or something that\`s on fire 🔥 (e.g. incident related), you can skip this step.
- [x] Read the [contribution guidelines](../CONTRIBUTING.md) 📖. This ensures your code follows our style guide and passes our tests.
- [ ] Update or add tests (if any source code was changed or added) 🧪
- [ ] Update the [included software doc](../included_software.md) (if you updated included software) 📄
- [ ] Update or add documentation (if features were changed or added) 📝
- [ ] Make sure the status checks below are successful ✅

**A picture of a cute animal (not mandatory, but encouraged)**
Here's my cat Binx!
<details>

image

</details>

### CODE REVIEW

**Comment by zackerydev** [comment]
Existing tests are passing - there doesn't seem to be any npm specific tests here so let me know if I need to add one

File: run-build-functions.sh [hunk 1]
```diff
@@ -184,6 +187,14 @@ run_npm() {
...
   fi
 
+  if [ -n "$EXPERIMENTAL_NPM_WORKSPACES_CACHING" ]
+  then
+    restore_node_modules "npm"
+  else
+    restore_cwd_cache node_modules "node modules"
+  fi
+
+
```
Comment (ehmicky) [inline run-build-functions.sh]: [sand] Small note: this file currently does not use double empty lines, so we might want to remove this empty line.

File: run-build-functions.sh [hunk 2]
```diff
@@ -184,6 +187,14 @@ run_npm() {
     fi
   fi
 
+  if [ -n "$EXPERIMENTAL_NPM_WORKSPACES_CACHING" ]
```
Comment (ehmicky) [inline run-build-functions.sh]: Thanks a lot for adding a feature flag!

File: run-build-functions.sh [hunk 3]
```diff
@@ -166,8 +171,6 @@ run_npm_set_temp() {
 }
 
 run_npm() {
-  restore_cwd_cache node_modules "node modules"
```
Comment (ehmicky) [inline run-build-functions.sh:169]: I am curious whether there was a reason this line was moved after `npm install -g npm`?
If not, we might want to keep the current place.

File: run-build-functions.sh [hunk 4]
```diff
@@ -86,40 +86,18 @@ install_deps() {
...
-  then
-    echo "Installing yarn at version $yarn_version"
-    rm -rf $HOME/.yarn
-    bash /usr/local/bin/yarn-installer.sh --version $yarn_version
-    mv $HOME/.yarn $NETLIFY_CACHE_DIR/yarn
-    export PATH=$NETLIFY_CACHE_DIR/yarn/bin:$PATH
-  fi
-
-
+restore_node_modules() {
```
Comment (ehmicky) [inline run-build-functions.sh:89]: Nice code re-use :+1:

Comment (zackerydev) [reply run-build-functions.sh:89]: Thanks!

File: run-build-functions.sh (hunk 3 above)
Comment (zackerydev) [reply run-build-functions.sh:169]: There wasn't I'm not sure why I moved it honestly 😅 

Addressed in 68778bb9f90782fdfcdb4ac953bd6ff835bc8b30

File: run-build-functions.sh (hunk 1 above)
Comment (zackerydev) [reply run-build-functions.sh]: Addressed in 68778bb9f90782fdfcdb4ac953bd6ff835bc8b30!

**Comment by zackerydev** [comment]
@ehmicky Sorry for the delay - comments addressed in 68778bb9f90782fdfcdb4ac953bd6ff835bc8b30!

**Comment by ehmicky** [comment]
Thanks a lot @zgriesinger, this is awesome! :tada: 

This is looking great.
We have a code freeze (for unrelated reasons) today and tomorrow, so I'll release this on Thursday. Thanks again!

**Review by ehmicky — APPROVED** [review]

**Comment by ehmicky** [comment]
We have extended our code freeze, so we will release this on Monday instead.

**Comment by zackerydev** [comment]
@ehmicky Any update on this?

**Comment by ehmicky** [comment]
Yes!
We extended our code freeze until yesterday. I am now going to release this. :+1:

**Comment by ehmicky** [comment]
This has been released!
@zgriesinger Could you please try it and confirm it works as you expected (after setting the `EXPERIMENTAL_NPM_WORKSPACES_CACHING` environment variable)? Thanks!

**Comment by zackerydev** [comment]
Just tried it @ehmicky 

Works great so far! Our build time on npm 8 workspaces actually hits cache now and this took our build times from around 10 minutes (due to never hitting cache and npm 7 not being workspace aware) to 3 minutes when hitting cache!

**Comment by ehmicky** [comment]
Oh wow, this is an awesome improvement, yay to caching! :tada: 

I think we're good with removing the `EXPERIMENTAL_NPM_WORKSPACES_CACHING` feature flag so everyone with npm workspaces can benefit from it, what do you think? If so, would you be interested in submitting a PR to remove it?

Thanks a lot again @zgriesinger.

**Comment by zackerydev** [comment]
@ehmicky sure I'd be happy to remove the feature flag!

I will give it a little bit of time to test it thoroughly then make the PR

**Comment by ehmicky** [comment]
One thing pointed out by @JGAntunes: it looks like the `yarn workspaces` command requires the root `package.json` to be `private: true`. This only applies to Yarn v1 (which is the version that's available by default in Netlify builds).

**Comment by ehmicky** [comment]
@zgriesinger how did your testing go?

**Comment by zackerydev** [comment]
@ehmicky Running into a few weird issues with _installing_ npm

I'm not sure what the deal is but my build logs are showing:
```
10:12:06 AM: npm workspaces detected
10:12:06 AM: Started restoring workspace apps/app/ui node modules
10:12:06 AM: Finished restoring workspace apps/app/ui node modules
10:12:06 AM: Started restoring workspace packages/react/components node modules
10:12:06 AM: Finished restoring workspace packages/react/components node modules
10:12:06 AM: Started restoring workspace packages/react/services node modules
10:12:06 AM: Finished restoring workspace packages/react/services node modules
10:12:06 AM: Started restoring workspace root node modules
10:12:06 AM: Finished restoring workspace root node modules
10:12:06 AM: npm WARN ignoring workspace config at /opt/build/repo/apps/app/ui/.npmrc
10:12:07 AM: npm WARN ignoring workspace config at /opt/build/repo/apps/app/ui/.npmrc
10:12:07 AM: Found npm version (8.5.2) that doesn't match expected (8.5.5)
Installing npm at version 8.5.5
10:12:07 AM: npm WARN ignoring workspace config at /opt/build/repo/apps/app/ui/.npmrc
10:12:07 AM: npm ERR! Workspaces not supported for global packages
10:12:07 AM: npm ERR! A complete log of this run can be found in:
10:12:07 AM: npm ERR!     /opt/buildhome/.npm/_logs/2022-03-21T15_12_07_298Z-debug-0.log
10:12:07 AM: Error installing NPM
```
So digging into that this week, trying to figure out what's changed since the successful builds

**Comment by ehmicky** [comment]
Oh nice catch @zgriesinger! Please let us know once you find the problem.

**Comment by zackerydev** [comment]
@ehmicky I think it was just bad cache, we've been running all of our workspaces sites on npm 8 and the new caching this week and things look good! I'll PR removing the flag later this week or early next!

**Comment by ehmicky** [comment]
Nice, thanks a lot @zgriesinger!

**Comment by ehmicky** [comment]
I opened a PR to remove the feature flag at https://github.com/netlify/build-image/pull/764

**Comment by zackerydev** [comment]
Awesome thank you @ehmicky I was just getting to it today!

**Comment by ehmicky** [comment]
This is now fully released for every user :tada:

**Pull request merged**